src/cpp/src/continuous_batching_impl.cpp

```diff
@@ -299,8 +305,38 @@ ContinuousBatchingPipeline::ContinuousBatchingImpl::generate(const std::vector<o
     }
     auto all_requests = m_awaiting_requests; // we need to store all requests to get results from them once generation has finished
 
-    bool continue_generation = true;
-    while (has_non_finished_requests() && continue_generation) {
+    std::atomic<bool> has_active_request = has_non_finished_requests();
+    GenerationHandle& generation = generations.at(0);
+
+    // create variables to make optimal thread-safe streaming
+    std::mutex mutex;
+    std::unique_lock lock(mutex);
+    std::condition_variable cv;
+
+    // define stream token lambda to use in `t_stream`
+    auto stream_tokens = [&generation, &streamer_ptr, &has_active_request, &cv, &lock]() {
+        while (!generation->is_dropped() && (has_active_request || streamer_ptr && generation->can_read())) {
+            // waiting for any tokens or request finishing
+            cv.wait(lock, [&generation, &has_active_request]{ return generation->can_read() || !has_active_request; });
+            if (streamer_ptr && generation->can_read()) {
+                std::unordered_map<uint64_t, GenerationOutput> token = generation->back();
+                for (const auto& gen_token : token.begin()->second.generated_ids) {
+                    if (streamer_ptr->put(gen_token)) {
+                        generation->drop();
+                        cv.notify_all();
```

Comment:
who is notified here? `cv.wait()` is used only in current thread, so it seems notification is not required.